core/deployment/src/main/java/io/quarkus/deployment/configuration/type/ConverterType.java

```diff
@@ -0,0 +1,112 @@
+package io.quarkus.deployment.configuration.type;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Field;
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+
+import io.quarkus.deployment.util.ReflectUtil;
+import io.quarkus.runtime.annotations.ConvertWith;
+import io.quarkus.runtime.annotations.DefaultConverter;
+import io.quarkus.runtime.configuration.HyphenateEnumConverter;
+
+/**
+ *
+ */
+public abstract class ConverterType {
+    ConverterType() {
+    }
+
+    public abstract Class<?> getLeafType();
+
+    public static ConverterType of(Field member) {
+        return of(member.getGenericType(), member);
+    }
+
+    public static ConverterType of(Parameter parameter) {
+        return of(parameter.getParameterizedType(), parameter);
+    }
+
+    public static ConverterType of(Type type, AnnotatedElement element) {
+        if (type instanceof GenericArrayType) {
+            GenericArrayType genericArrayType = (GenericArrayType) type;
+            return new ArrayOf(of(genericArrayType.getGenericComponentType(), element));
+        } else if (type instanceof Class<?>) {
+            // simple type
+            Class<?> clazz = (Class<?>) type;
+            if (clazz.isArray()) {
+                return new ArrayOf(of(clazz.getComponentType(), element));
+            }
+            ConvertWith convertWith = element.getAnnotation(ConvertWith.class);
+            Leaf leaf;
+            if (convertWith == null && element.getAnnotation(DefaultConverter.class) == null && clazz.isEnum()) {
+                // use our hyphenated converter by default
+                leaf = new Leaf(clazz, HyphenateEnumConverter.class);
+            } else {
+                leaf = new Leaf(clazz, convertWith == null ? null : convertWith.value());
+            }
+            // vvv todo: add validations here vvv
```

Comment:
Should this be addressed?

Reply:
Later.  It has to do with #2489 - this is the point where support for that would be inserted.